package/MDAnalysis/analysis/leaflet.py

```diff
@@ -322,3 +322,255 @@ def optimize_cutoff(universe, select, dmin=10.0, dmax=20.0, step=0.5,
+        self.residues = universe.select_atoms(select_residues).residues
+        self.n_residues = len(self.residues)
+        self.headgroups = self.residues.atoms.select_atoms(select_headgroup)
+        if n_leaflets < 1:
+            raise ValueError('Must have at least one leaflet')
+        self.n_leaflets = n_leaflets
+        self.delta = delta
+        self.compute_headgroup_only = compute_headgroup_only
+        self.cutoff = enrichment_cutoff
+        self.leaflets = []
+        self.leaflets_summary = []
+
+        # process this a bit to remove errors like "alt_locs"
+        attrname = _TOPOLOGY_ATTRNAMES[count_by_attr.lower().replace('_', '')]
+        self.attrname = _TOPOLOGY_ATTRS[attrname].attrname
+
+    def _prepare(self):
+        try:
+            import sklearn.cluster as skc
+        except ImportError:
+            raise ImportError('scikit-learn is required to use this analysis '
+                              'but is not installed. Install it with `conda '
+                              'install scikit-learn` or `pip install '
+                              'scikit-learn`.') from None
+
+        if self.n_leaflets > 1:
+            # determine leaflets by clustering center-of-geometry
+            coms = self.headgroups.center_of_geometry(compound='residues')
+            pdist = distances.distance_array(coms, coms, box=self.box)
+            psim = np.exp(-pdist**2/(2*(self.delta**2)))
+            sc = skc.SpectralClustering(n_clusters=self.n_leaflets,
+                                        affinity='precomputed_nearest_neighbors')
+            clusters = sc.fit_predict(pdist)
+            leaflets = []
+            for i in range(self.n_leaflets):
+                leaflets.append([])
+            for res, i in zip(self.residues, clusters):
+                leaflets[i].append(res)
+            self.leaflet_residues = lres = [np.array(x).sum() 
+                                            for x in leaflets]
+            self.leaflet_residues.sort(key=lambda x: x.center_of_geometry()[-1],
+                                       reverse=True)
+            self.leaflet_headgroups = [(self.headgroups & r.atoms) 
+                                       for r in lres]
+        else:
+            self.leaflet_residues = [self.residues]
+            self.leaflet_headgroups = [self.headgroups]
+        
+        if not self.compute_headgroup_only:
+            self.leaflet_headgroups = [h.residues.atoms
+                                       for h in self.leaflet_headgroups]
+
+        # set up results
+        self.residue_counts = [np.zeros((len(r), self.n_frames)) for r in lres]
+        self.total_counts = np.zeros((self.n_leaflets, self.n_frames))
+        self.leaflet_ids = ids = [getattr(r, self.attrname) for r in lres]
+        self.leaflets = [dict.fromkeys(np.unique(r)) for r in ids]
+        self.leaflets_summary = [dict.fromkeys(np.unique(r)) for r in ids]
+
+    def _single_frame(self):
```

Comment:
You don't rebuild leaflet groups at every step. For really long (or cg) sims you might have lipid flip-flops and you start assigning lipids wrongly.  At least document.

One could also have a setting that leaflets are recomputed every n steps.

Reply:
> One could also have a setting that leaflets are recomputed every n steps.

Yeah, I did consider that. I didn't want to cluster on every frame by default. I was thinking instead of computing distance to the center of geometry of the different lipids, which would be less computationally expensive. I would have to change up how I structure the results containers, though, as currently they're numpy arrays where each element represents a lipid :/

Reply:
For right now I would just note it in the documentation as a potential limitation. If there’s sufficient need then whoever has the greatest need can work on improving it.

Reply:
Yes, on the other hand, that's probably me by Tuesday 😄

Reply:
I got to play with the original code from the reference paper, we often observed flip-flop, especially just near to the proteins. It is especially true when looking at small lipids or, worst, sterols. Accounting for flip-flop is the difficult part in this analysis.